pkg/cmd/api/api.go

```diff
@@ -0,0 +1,190 @@
+package api
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
+
+type ApiOptions struct {
+	IO *iostreams.IOStreams
+
+	RequestMethod       string
+	RequestMethodPassed bool
+	RequestPath         string
+	MagicFields         []string
+	RawFields           []string
+	RequestHeaders      []string
+	ShowResponseHeaders bool
+
+	HttpClient func() (*http.Client, error)
+}
+
+func NewCmdApi(f *cmdutil.Factory, runF func(*ApiOptions) error) *cobra.Command {
+	opts := ApiOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "api <endpoint>",
+		Short: "Make an authenticated GitHub API request",
+		Long: `Makes an authenticated HTTP request to the GitHub API and prints the response.
+
+The <endpoint> argument should either be a path of a GitHub API v3 endpoint, or
+"graphql" to access the GitHub API v4.
+
+The default HTTP request method is "GET" normally and "POST" if any parameters
+were added. Override the method with '--method'.
+
+Pass one or more '--raw-field' values in "<key>=<value>" format to add
+JSON-encoded string parameters to the POST body.
+
+The '--field' flag behaves like '--raw-field' with magic type conversion based
+on the format of the value:
+
+- literal values "true", "false", "null", and integer numbers get converted to
+  appropriate JSON types;
+- if the value starts with "@", the rest of the value is interpreted as a
+  filename to read the value from. Pass "-" to read from standard input.
+`,
+		Args: cobra.ExactArgs(1),
+		RunE: func(c *cobra.Command, args []string) error {
+			opts.RequestPath = args[0]
+			opts.RequestMethodPassed = c.Flags().Changed("method")
+
+			if runF != nil {
+				return runF(&opts)
+			}
+			return apiRun(&opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.RequestMethod, "method", "X", "GET", "The HTTP method for the request")
+	cmd.Flags().StringArrayVarP(&opts.MagicFields, "field", "F", nil, "Add a parameter of inferred type")
+	cmd.Flags().StringArrayVarP(&opts.RawFields, "raw-field", "f", nil, "Add a string parameter")
+	cmd.Flags().StringArrayVarP(&opts.RequestHeaders, "header", "H", nil, "Add an additional HTTP request header")
+	cmd.Flags().BoolVarP(&opts.ShowResponseHeaders, "include", "i", false, "Include HTTP response headers in the output")
+	return cmd
+}
+
+func apiRun(opts *ApiOptions) error {
+	params, err := parseFields(opts)
+	if err != nil {
+		return err
+	}
+
+	method := opts.RequestMethod
+	if len(params) > 0 && !opts.RequestMethodPassed {
+		method = "POST"
+	}
```

Comment:
It seems a little weird that opts has RequestMethod and RequestMethodPassed. Moving this logic into the RunE function and assuming RequestMethod is always set correctly is also a little weird, but I think it might be easier to understand.

Reply:
That can also work! I can swing both ways with this approach in particular; I just wanted to set some precedent to communicating via opts struct that a certain flag has been passed or not, since we use that information in some other commands as well. I agree that since `NewCmdApi()` is only concerned with setting up the `cobra.Command` object and defining flag parsing, it should also handle as much of the flag processing logic as it can before dispatching to the `apiRun()` command.